### PR TITLE
Assignment7+8 linq Michael Moran @mmoran314 Ben Fristad @BenFristadSchool

### PR DESCRIPTION
**Pull request by mmoran314**
Assignment7+8 linq Michael Moran @mmoran314 Ben Fristad @BenFristadSchool

### CODE REVIEW

**Comment by jacobnewcomb** [comment]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Currently set to “Do not Copy”

Using LINQ, skip the first row in the People.csv. ✔

Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

Implement IEnumerable GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Although the functionality seems correct the Assert.AreEqual test Fails. Loop through each of them validating if they are equal in the hardcoded test

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Although there is a hardcoded test, it fails - loop through each individual value

Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Should not use for loop to loop through it, utilize linq queries instead

Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔

Given the array, consider using string.Join to combine the list into a single string. ✔

Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable to return all the items in the "circle" of items. ✔
Not Tested

Add an IEnumberable ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. ✔
Test Fails

Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by VeraKorchemnaya** [comment]
I will review.

**Review by VeraKorchemnaya — COMMENTED** [review]
## Instructions

**Throughout, consider using the `System.Linq.Enumerable` methods `Zip`, `Count`, `Sort` and `Contains` methods for testing collections.**. (Preferably avoid using `Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert` although that might be easier, to get a firmer grasp on additional LINQ API.)

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. (Sort the addresses first then select). ❌ (You need to also sort by city and zip)
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ✔

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma-separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result. ❌ (It says recommended and not required but I would probably still do it)

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`. ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version) ✔
- All of the above should be unit tested ❌ (Missing Node GetEnumerator() test)
- Choose simplicity over complexity ✔ (Take a look at my comment on 6)

## Comments

- A lot of commented code that I would remove and some comments that seem unnecessary
- Take a peek at the Error List because you are getting some messages that are valid
  - Some typos to fix
  - Some code can be simplify

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        //Act
+        IEnumerable<string> statesSorted = states.Distinct().ToImmutableSortedSet();
+
+        //Assert
+        for (int i = 0; i < statesSorted.Count(); i++) 
+        {
+            Assert.AreEqual<string>(statesSorted.ElementAt(i), expected.ElementAt(i));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mmoran314) [reply Assignment/Assignment.Tests/SampleDataTests.cs:72]: Yeah we were a little confused on this one, I guess when it was mentioned in class and why we were using a hardcoded list anyway makes me believe this is the way and by using the same logic that the method uses. I could be wrong but I think I will stick with this.

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,88 @@
...
+            
 
         // 4.
-        public IEnumerable<IPerson> People => throw new NotImplementedException();
+        public IEnumerable<IPerson> People 
+        {
+            get
+            {
+                List<Person> persons = new List<Person>();
+                foreach (string row in CsvRows)
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:35]: You should use select here, not foreach

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,88 @@
...
 
+
+
+        // 5.
+        public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(Predicate<string> filter)
+        {
+            
+            List<IPerson> filtered = People.Where(person => filter(person.EmailAddress)).ToList();
+            List<(string, string)> personList = new List<(string, string)>();
+            foreach (IPerson person in filtered)
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:64]: Same here

**Pull request closed without merging**